### ISSUE TITLE
perl 5.6.1 with Intel's icc on RedHat Linux 7.2

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4591]
Migrated from [rt.perl.org#7916](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7916) (status was 'resolved')

Searchable as RT7916$

**Comment by p5pRT** [comment]
### From cfriv@yahoo.com

<details><summary>Inline Patch</summary>

```diff
--- ./backup/Errno_pm.PL	Fri Nov 16 09:44:51 2001
+++ Errno_pm.PL	Fri Nov 16 09:49:22 2001
@@ -98,7 +98,7 @@
     } elsif ($Config{archname} eq 'epoc') {
 	# Watch out for cross compiling for EPOC (usually done on linux)
 	$file{'/usr/local/epoc/include/libc/sys/errno.h'} = 1;
-    } elsif ($^O eq 'linux') {
+    } elsif ($^O eq 'linux' && $Config{cc} ne "icc" ) {
 	# Some Linuxes have weird errno.hs which generate
 	# no #file or #line directives
 	$file{'/usr/include/errno.h'} = 1;
```
</details>

**Comment by p5pRT** [comment]
### From @jhi

&nbsp;    http&#8203;://www\.iki\.fi/jhi/perl@&#8203;13054\.tgz

\-\- 
$jhi\+\+; \# http&#8203;://www\.iki\.fi/jhi/
&nbsp;        \# There is this special biologist word we use for 'stable'\.
&nbsp;        \# It is 'dead'\. \-\- Jack Cohen

**Comment by p5pRT** [comment]
### From @smpeters

> \[cfriv@&#8203;yahoo\.com \- Fri Nov 16 01&#8203;:32&#8203;:46 2001\]&#8203;:
> 
> \-\-\- \./backup/Errno\_pm\.PL    Fri Nov 16 09&#8203;:44&#8203;:51 2001
> \+\+\+ Errno\_pm\.PL    Fri Nov 16 09&#8203;:49&#8203;:22 2001
> @&#8203;@&#8203; \-98\,7 \+98\,7 @&#8203;@&#8203;
>      \} elsif \($Config\{archname\} eq 'epoc'\) \{
>      \# Watch out for cross compiling for EPOC \(usually done on linux\)
>      $file\{'/usr/local/epoc/include/libc/sys/errno\.h'\} = 1;
> \-    \} elsif \($^O eq 'linux'\) \{
> \+    \} elsif \($^O eq 'linux' && $Config\{cc\} ne "icc" \) \{
>      \# Some Linuxes have weird errno\.hs which generate
>      \# no \#file or \#line directives
>      $file\{'/usr/include/errno\.h'\} = 1;
> 
> 

The change implemented in change \#13053 appears OK for older icc's\. 
Unfortunately\, newer icc's masquerade as gcc\, with unfortunate
side\-effects\.  I'll take a closer look at that change this evening\.

**Comment by p5pRT** [comment]
### From @nwc10

On Mon\, May 01\, 2006 at 08&#8203;:52&#8203;:09AM \-0700\, Steve Peters via RT wrote&#8203;:
> > \[cfriv@&#8203;yahoo\.com \- Fri Nov 16 01&#8203;:32&#8203;:46 2001\]&#8203;:
> > 
> > \-\-\- \./backup/Errno\_pm\.PL    Fri Nov 16 09&#8203;:44&#8203;:51 2001
> > \+\+\+ Errno\_pm\.PL    Fri Nov 16 09&#8203;:49&#8203;:22 2001
> > @&#8203;@&#8203; \-98\,7 \+98\,7 @&#8203;@&#8203;
> >      \} elsif \($Config\{archname\} eq 'epoc'\) \{
> >      \# Watch out for cross compiling for EPOC \(usually done on linux\)
> >      $file\{'/usr/local/epoc/include/libc/sys/errno\.h'\} = 1;
> > \-    \} elsif \($^O eq 'linux'\) \{
> > \+    \} elsif \($^O eq 'linux' && $Config\{cc\} ne "icc" \) \{
> >      \# Some Linuxes have weird errno\.hs which generate
> >      \# no \#file or \#line directives
> >      $file\{'/usr/include/errno\.h'\} = 1;
> > 
> > 
> 
> The change implemented in change \#13053 appears OK for older icc's\. 
> Unfortunately\, newer icc's masquerade as gcc\, with unfortunate
> side\-effects\.  I'll take a closer look at that change this evening\.

Well\, it's fairly easy to write a probe\. For example\, compile this with
optimisation&#8203;:

\#include \<stdlib\.h>
\#include \<stdio\.h>

int main \(int argc\, char \*\*argv\) \{
&nbsp;  double nan = atof\("NaN"\);

&nbsp;  puts\(nan == nan ? "icc" : "gcc"\);
&nbsp;  return 0;
\}


If it gives the wrong answer\, it's icc\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @smpeters

On Mon\, May 01\, 2006 at 05&#8203;:40&#8203;:48PM \+0100\, Nicholas Clark wrote&#8203;:
> On Mon\, May 01\, 2006 at 08&#8203;:52&#8203;:09AM \-0700\, Steve Peters via RT wrote&#8203;:
> > > \[cfriv@&#8203;yahoo\.com \- Fri Nov 16 01&#8203;:32&#8203;:46 2001\]&#8203;:
> > > 
> > > \-\-\- \./backup/Errno\_pm\.PL    Fri Nov 16 09&#8203;:44&#8203;:51 2001
> > > \+\+\+ Errno\_pm\.PL    Fri Nov 16 09&#8203;:49&#8203;:22 2001
> > > @&#8203;@&#8203; \-98\,7 \+98\,7 @&#8203;@&#8203;
> > >      \} elsif \($Config\{archname\} eq 'epoc'\) \{
> > >      \# Watch out for cross compiling for EPOC \(usually done on linux\)
> > >      $file\{'/usr/local/epoc/include/libc/sys/errno\.h'\} = 1;
> > > \-    \} elsif \($^O eq 'linux'\) \{
> > > \+    \} elsif \($^O eq 'linux' && $Config\{cc\} ne "icc" \) \{
> > >      \# Some Linuxes have weird errno\.hs which generate
> > >      \# no \#file or \#line directives
> > >      $file\{'/usr/include/errno\.h'\} = 1;
> > > 
> > > 
> > 
> > The change implemented in change \#13053 appears OK for older icc's\. 
> > Unfortunately\, newer icc's masquerade as gcc\, with unfortunate
> > side\-effects\.  I'll take a closer look at that change this evening\.
> 
> Well\, it's fairly easy to write a probe\. For example\, compile this with
> optimisation&#8203;:
> 
> \#include \<stdlib\.h>
> \#include \<stdio\.h>
> 
> int main \(int argc\, char \*\*argv\) \{
>   double nan = atof\("NaN"\);
> 
>   puts\(nan == nan ? "icc" : "gcc"\);
>   return 0;
> \}
> 
> 
> If it gives the wrong answer\, it's icc\.
> 

The following should be enough\.\.\.

Change 28043 by stevep@&#8203;stevep\-kirk on 2006/05/02 00&#8203;:23&#8203;:05

&nbsp;    Fix for RT \#7916&#8203;: perl 5\.6\.1 with Intel's icc on RedHat 
&nbsp;    Linux 7\.2\.  The original fix\, change \#13053\, added a check
&nbsp;    for $Config\{gccversion\} which is now populated by newer
&nbsp;    Intel compilers\.  This additional change checks to see
&nbsp;    if Intel is in $Config\{gccversion\}\.  If it is\, its not
&nbsp;    a gcc\.

Affected files \.\.\.

\.\.\. //depot/perl/ext/Errno/Errno\_pm\.PL\#47 edit

Differences \.\.\.

==== //depot/perl/ext/Errno/Errno\_pm\.PL\#47 \(text\) ====

@&#8203;@&#8203; \-141\,7 \+141\,9 @&#8203;@&#8203;
&nbsp;          $file\{$1\} = 1;
&nbsp;     \}
&nbsp;     \} elsif \($^O eq 'linux' &&
\-         $Config\{gccversion\} ne '' \# might be using\, say\, Intel's icc
\+         \($Config\{gccversion\} ne '' && 
\+          $Config\{gccversion\} \!~ /intel/i
\+          \# might be using\, say\, Intel's icc
&nbsp;          \) \{
&nbsp;     \# Some Linuxes have weird errno\.hs which generate
&nbsp;     \# no \#file or \#line directives

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
